src/ocean/io/select/EpollSelectDispatcher.d

```diff
@@ -740,10 +787,17 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
         bool caller_work_pending = false;
 
         if (select_cycle_hook !is null)
-            caller_work_pending = select_cycle_hook();
+            select_cycle_hooks ~= select_cycle_hook;
+
+        foreach (hook; select_cycle_hooks)
+        {
+            caller_work_pending |= hook();
```

Comment:
`hook()` may call `cycleHook` and/or `removeCycleHook`. This leads to complications which need to be addressed:
1. `cycleHook` increases the length of the `select_cycle_hooks` array. This can cause buffer relocation, which would turn the array iterating over into a dangling slice. Using a `for` loop instead should be safe AFAICS.
2. `removeCycleHook` changes the order of the elements in `select_cycle_hooks`, it moves the last to the position of the removed element. If this happens during this iteration and the removed hook has already been called, the previously last hook in the array will be skipped. I can think of two ways to prevent this from happening:
    - Do not move array elements in `removeCycleHook`, instead set the deleted element to `null`.
    - Use a linked list rather than an array for `select_cycle_hooks`.

Reply:
Yeah this is one thing that should be addressed by https://github.com/sociomantic-tsunami/ocean/pull/666